File: Chapter 1/TwoPointer/SquaresofSorted.js
var nums = [-4,-1,0,3,10]
// Output: [0,1,9,16,100]

// const squareSorted = (nums) => { 
//   return nums.map(num => num * num).sort((a, b) => a -b) 
// }

const sorted = (nums) => { 
  let left = 0;
  let right = nums.length - 1
  let pos = nums.length -1;
  let array = new Array(nums.length)

  while(left <= right){ 
    let lefty = nums[left] * nums[left]
    let righty = nums[right] * nums[right]

    if(lefty > righty){ 
      array[pos] = lefty 
      left++
    } else { 
      array[pos] = righty
      right--
    }
    pos--
  }
  return array
}

console.log(sorted(nums))

File: 2026/January-5-2026/SquaresOfSorted.py
# Solution
# Given an integer array nums sorted in non-decreasing order, return an array of the squares of each number sorted in non-decreasing order.

 

# Example 1:

# Input: nums = [-4,-1,0,3,10]
# Output: [0,1,9,16,100]

from typing import List

nums = [-4,-1,0,3,10]

def squares_of_sorted_array(nums: List[int]) -> List[int]:

  for i in range(len(nums)):
    nums[i] = nums[i]**2
  nums.sort()

  return nums

print(squares_of_sorted_array(nums)) 


def refactored_sq_sorted(nums: List[int]) -> List[int]: 
  n = len(nums)
  result = [0] * n
  left = 0 
  right = n - 1
  pos = n - 1

  while left <= right: 
    if abs(nums[left]) < abs(nums[right]): 
      result[pos] = nums[left]**2
      left += 1
    else: 
      result[pos] = nums[right]**2
      right -= 1
    pos -=1 
  return nums

print(refactored_sq_sorted(nums))